# DESASTRES NATURALES ENTRE 1970-2021

La Base de Datos Internacional sobre Desastres (EM-DAT o The International Disaster Database) contiene datos sobre la ocurrencia y el impacto de más de 27.000 desastres masivos en todo el mundo desde 1900 hasta la actualidad. La base de datos se compila a partir de diversas fuentes, como agencias de la ONU, organizaciones no gubernamentales, compañías de reaseguros, institutos de investigación y agencias de prensa. El Centro de Investigación sobre la Epidemiología de los Desastres (CRED) distribuye los datos en acceso abierto para uso no comercial.

De este gran conjunto de datos, hemos tomado datos entre 1970 y 2021, para ver los desastres naturales de todos los países del mundo en éste período.

## ANÁLISIS EXPLORATORIO DE LOS DATOS: INFORME EDA

Las columnas que encontramos en el DataFrame del archivo "1970-2021_natural_disasters-emdat_data.csv" son:

- Dis No – Código único de desastre.

- Year – Año del desastre.

- Seq – Número de secuencia en el año según la base de datos EM-DAT. 

- Glide – Identificador GLIDE (Global Unique Disaster Identifier). 

- Disaster Group – Grupo principal (ej. Natural, Technological).

- Disaster Subgroup – Subgrupo (ej. Meteorological, Hydrological).

- Disaster Type – Tipo (ej. Storm, Flood).

- Disaster Subtype – Subtipo (ej. Tropical cyclone).

- Disaster Subsubtype – Sub-subtipo (si aplica). 

- Event Name – Nombre del evento (si existe).

- Country – País afectado.

- ISO – Código ISO del país. 

- Region – Región (ej. Americas, Asia). 

- Continent – Continente.

- Location – Lugar específico.

- Origin – Origen del desastre (ej. Natural, Industrial).

- Associated Dis – Desastres asociados (1). 

- Associated Dis2 – Desastres asociados (2). 

- OFDA Response – Respuesta de la Office of U.S. Foreign Disaster Assistance.

- Appeal – Si hubo apelación internacional.

- Declaration – Declaración oficial de desastre.

- Aid Contribution – Contribución de ayuda (si se registró). 

- Dis Mag Value – Valor de magnitud (ej. magnitud sísmica).

- Dis Mag Scale – Escala de magnitud (ej. Richter, Mercalli).

- Latitude – Latitud del evento.

- Longitude – Longitud del evento.

- Local Time – Hora local del inicio del desastre.

- River Basin – Cuenca hidrográfica (si aplica).

- Start Year – Año de inicio.

- Start Month – Mes de inicio.

- Start Day – Día de inicio.

- End Year – Año de finalización. 

- End Month – Mes de finalización.

- End Day – Día de finalización.

- Total Deaths – Número total de fallecidos.

- No Injured – Número de heridos.

- No Affected – Número de afectados.

- No Homeless – Número de personas sin hogar.

- Total Affected – Total de personas afectadas (suma de afectados, heridos, sin hogar).

- Reconstruction Costs ('000 US$) – Costos de reconstrucción (en miles de USD). 

- Insured Damages ('000 US$) – Daños asegurados (en miles de USD).

- Total Damages ('000 US$) – Daños totales (en miles de USD). 

- CPI – Índice de precios al consumidor (para ajuste económico). 

- Adm Level – Nivel administrativo (ej. nacional, regional).

- Admin1 Code – Código de primera división administrativa (ej. estado, provincia).

- Admin2 Code – Código de segunda división administrativa (ej. municipio).

- Geo Locations – Ubicaciones geográficas adicionales.

Este conjunto de datos proporciona información clara sobre eventos de desastre natural a nivel regional, lo que lo hace útil para el mapeo, los estudios de riesgo y la planificación de respuestas a desastres.

## INFORME DE TRANSFORMACIÓN DE LOS DATOS

Incluye la limpieza de datos, la normalización, la conversión de tipos de datos y revisión específica de datos nulos. Las transformaciones se han realizado mediante funciones de Python aplicados a los datos extraídos.

In [1]:
# Importamos las librerias a utilizar
import pandas as pd
import numpy as np


# Configuración para poder visualizar todas las columnas del DataFrame
# -----------------------------------------------------------------------
pd.set_option('display.max_columns', None)

In [2]:
df = pd.read_csv("1970-2021_natural_disasters-emdat_data.csv", index_col=0)
df.head()

Year  Seq Glide Disaster Group Disaster Subgroup Disaster Type  \
Dis No                                                                          
1970-0013-ARG  1970   13   NaN        Natural      Hydrological         Flood   
1970-0109-AUS  1970  109   NaN        Natural    Meteorological         Storm   
1970-0044-BEN  1970   44   NaN        Natural      Hydrological         Flood   
1970-0063-BGD  1970   63   NaN        Natural    Meteorological         Storm   
1970-0026-BGD  1970   26   NaN        Natural    Meteorological         Storm   

               Disaster Subtype Disaster Subsubtype Event Name     Country  \
Dis No                                                                       
1970-0013-ARG               NaN                 NaN        NaN   Argentina   
1970-0109-AUS  Tropical cyclone                 NaN        Ada   Australia   
1970-0044-BEN               NaN                 NaN        NaN       Benin   
1970-0063-BGD  Tropical cyclone                 NaN        NaN  Bangladesh   
1970-0026-BGD               NaN                 NaN        NaN  Bangladesh   

               ISO                     Region Continent            Location  \
Dis No                                                                        
1970-0013-ARG  ARG              South America  Americas             Mendoza   
1970-0109-AUS  AUS  Australia and New Zealand   Oceania          Queensland   
1970-0044-BEN  BEN             Western Africa    Africa      Atacora region   
1970-0063-BGD  BGD              Southern Asia      Asia  Khulna, Chittagong   
1970-0026-BGD  BGD              Southern Asia      Asia                 NaN   

              Origin Associated Dis Associated Dis2 OFDA Response Appeal  \
Dis No                                                                     
1970-0013-ARG    NaN            NaN             NaN           NaN    NaN   
1970-0109-AUS    NaN            NaN             NaN           NaN    NaN   
1970-0044-BEN    NaN            NaN             NaN           Yes    NaN   
1970-0063-BGD    NaN            NaN             NaN           Yes    NaN   
1970-0026-BGD    NaN            NaN             NaN           NaN    NaN   

              Declaration  Aid Contribution  Dis Mag Value Dis Mag Scale  \
Dis No                                                                     
1970-0013-ARG         NaN               NaN            NaN           Km2   
1970-0109-AUS         NaN               NaN            NaN           Kph   
1970-0044-BEN         NaN               NaN            NaN           Km2   
1970-0063-BGD         NaN               NaN            NaN           Kph   
1970-0026-BGD         NaN               NaN            NaN           Kph   

              Latitude Longitude Local Time River Basin  Start Year  \
Dis No                                                                
1970-0013-ARG      NaN       NaN        NaN         NaN        1970   
1970-0109-AUS      NaN       NaN        NaN         NaN        1970   
1970-0044-BEN      NaN       NaN        NaN         NaN        1970   
1970-0063-BGD      NaN       NaN        NaN         NaN        1970   
1970-0026-BGD      NaN       NaN        NaN         NaN        1970   

               Start Month  Start Day  End Year  End Month  End Day  \
Dis No                                                                
1970-0013-ARG          1.0        4.0      1970        1.0      4.0   
1970-0109-AUS          1.0        NaN      1970        1.0      NaN   
1970-0044-BEN          9.0        NaN      1970        9.0      NaN   
1970-0063-BGD         11.0       12.0      1970       11.0     12.0   
1970-0026-BGD          4.0       13.0      1970        4.0     13.0   

               Total Deaths  No Injured  No Affected  No Homeless  \
Dis No                                                              
1970-0013-ARG          36.0         NaN          NaN          NaN   
1970-0109-AUS          13.0         NaN          NaN          NaN   
1970-004

In [3]:
df = df.reset_index()

In [4]:
df.head(2)

Dis No  Year  Seq Glide Disaster Group Disaster Subgroup  \
0  1970-0013-ARG  1970   13   NaN        Natural      Hydrological   
1  1970-0109-AUS  1970  109   NaN        Natural    Meteorological   

  Disaster Type  Disaster Subtype Disaster Subsubtype Event Name    Country  \
0         Flood               NaN                 NaN        NaN  Argentina   
1         Storm  Tropical cyclone                 NaN        Ada  Australia   

   ISO                     Region Continent    Location Origin Associated Dis  \
0  ARG              South America  Americas     Mendoza    NaN            NaN   
1  AUS  Australia and New Zealand   Oceania  Queensland    NaN            NaN   

  Associated Dis2 OFDA Response Appeal Declaration  Aid Contribution  \
0             NaN           NaN    NaN         NaN               NaN   
1             NaN           NaN    NaN         NaN               NaN   

   Dis Mag Value Dis Mag Scale Latitude Longitude Local Time River Basin  \
0            NaN           Km2      NaN       NaN        NaN         NaN   
1            NaN           Kph      NaN       NaN        NaN         NaN   

   Start Year  Start Month  Start Day  End Year  End Month  End Day  \
0        1970          1.0        4.0      1970        1.0      4.0   
1        1970          1.0        NaN      1970        1.0      NaN   

   Total Deaths  No Injured  No Affected  No Homeless  Total Affected  \
0          36.0         NaN          NaN          NaN             NaN   
1          13.0         NaN          NaN          NaN             NaN   

   Reconstruction Costs ('000 US$)  Insured Damages ('000 US$)  \
0                              NaN                         NaN   
1                              NaN                         NaN   

   Total Damages ('000 US$)        CPI Adm Level Admin1 Code Admin2 Code  \
0                   25000.0  15.001282       NaN         NaN         NaN   
1                   72475.0  15.001282       NaN         NaN         NaN   

  Geo Locations  
0           NaN  
1           NaN

In [5]:
print(f"El número de filas que tenemos es {df.shape[0]}, y el número de columnas es {df.shape[1]}")

El número de filas que tenemos es 14644, y el número de columnas es 47


Vamos a realizar la transformación columna por columna.

#### COLUMNA Dis No: Código único de desastre

- Índice 1
- Datos de tipo objeto (string)
- Incluye el año, Seq y ISO

In [6]:
df['Dis No'].value_counts()

Dis No
1970-0013-ARG    1
1970-0109-AUS    1
1970-0044-BEN    1
1970-0063-BGD    1
1970-0026-BGD    1
                ..
2021-0449-YEM    1
2021-0075-ZAF    1
2021-0599-COD    1
2021-0020-SRB    1
2021-0481-SSD    1
Name: count, Length: 14644, dtype: int64

#### COLUMNA Year: Año del desastre

- Índice 2
- Pensamos dejarla si conseguimos calcular la Duration para facilitar visualizaciones 

In [7]:
df['Year'].sample()

9798    2009
Name: Year, dtype: int64

#### COLUMNA Seq – Número de secuencia en el año según la base de datos EM-DAT.
- Índice 3
- La eliminamos ya que no aporta información extra al Dis No

#### COLUMNA Glide: Global Unique Disaster Identifier

- Índice 4
- 89% de nulos y datos sin valor. La eliminamos.

#### COLUMNA Disaster Group – Grupo principal (ej. Natural, Technological)

- Índice 5
- Datos de tipo objeto (string)
- No hay valores nulos, pero todos los valores son "Natural". Eliminamos la columna.

In [8]:
df['Disaster Group'].value_counts()

Disaster Group
Natural    14644
Name: count, dtype: int64

#### COLUMNA Disaster Subgroup – Subgrupo (ej. Meteorological, Hydrological)

- Índice 6
- Datos de tipo objeto (string)
- No hay valores nulos, pero sí observamos categorías del desastre.

In [9]:
df['Disaster Subgroup'].value_counts()

Disaster Subgroup
Hydrological         5985
Meteorological       4589
Biological           1530
Geophysical          1400
Climatological       1139
Extra-terrestrial       1
Name: count, dtype: int64

#### COLUMNA Disaster Type – Tipo (ej. Storm, Flood)

- Índice 7
- Datos de tipo objeto (string)
- No hay valores nulos, encontramos valores concretos sobre el tipo de desastre

In [10]:
df['Disaster Type'].value_counts()

Disaster Type
Flood                    5272
Storm                    4005
Epidemic                 1436
Earthquake               1137
Landslide                 713
Drought                   685
Extreme temperature       584
Wildfire                  452
Volcanic activity         222
Insect infestation         93
Mass movement (dry)        41
Glacial lake outburst       2
Impact                      1
Animal accident             1
Name: count, dtype: int64

#### COLUMNA Disaster Subtype – Subtipo (ej. Tropical cyclone)

- Índice 8
- Datos de tipo objeto (string)
- 2747 nulos - Repetimos información de "Disaster Type" para conservar los que sí tenemos info.

In [11]:
df["Disaster Subtype"] = df["Disaster Subtype"].fillna(df["Disaster Type"])

In [12]:
print("Nulos después:", df["Disaster Subtype"].isna().sum())

Nulos después: 0


#### COLUMNA Disaster Subsubtype – Sub-subtipo (si aplica)

- índice 9
- 93% de nulos, la eliminamos. Todos los subtypes tienen subsubtype, no nos ayuda para resolver los nulos de la columna anterior.

#### COLUMNA Event Name – Nombre del evento (si existe)

- Índice 10
- 75% de nulos - DE MOMENTO DEJAMOS como NAN

- Todos los Disaster Type tienen al menos un nulo en Event Name

* La mayoría provienen de "Flood" - 5261, siguiendo "Storm" - 1840 y "Earthquake" - 1136
* No la eliminamos, de las 3645 filas que tenemos nombre posteriormente podemos sacar info/gráficas.  

In [13]:
df['Event Name'].isnull().sum()

np.int64(10999)

In [14]:
nulos_por_type = (
    df.groupby("Disaster Type")["Event Name"]
      .apply(lambda x: x.isna().sum())
      .reset_index(name="Nombre nulo")
)

nulos_por_type

Disaster Type  Nombre nulo
0         Animal accident            1
1                 Drought          685
2              Earthquake         1136
3                Epidemic          268
4    Extreme temperature           583
5                   Flood         5261
6   Glacial lake outburst            2
7                  Impact            1
8      Insect infestation           36
9               Landslide          711
10    Mass movement (dry)           40
11                  Storm         1840
12      Volcanic activity            7
13               Wildfire          428

#### COLUMNA Country – País afectado

- Índice 11
- Datos de tipo objeto (string)
- No disponemos de valores nulos.
- El país que encabeza nuestra lista es USA, seguido de China.
- Filtrado realizado para evitar reiteraciones y algunos nombres antiguos resueltos (Checoslovaquia - Rep Checa). URSS conservada ya que no abarcaba sólo Rusia.

In [15]:
# Diccionario de normalización de países
replace_dict = {
    # Congo
    "Congo (the Democratic Republic of the)": "Democratic Republic of the Congo",
    "Congo (the)": "Republic of the Congo",

    # Czechia / Checoslovaquia
    "Czech Republic (the)": "Czech Republic",
    "Czechoslovakia": "Czech Republic",  # ya no existe, se fusiona con Czechia

    # Alemania
    "Germany Fed Rep": "Germany",
    "Germany Dem Rep": "Germany",

    # Guinea
    "Guinea-Bissau": "Guinea-Bissau",
    "Guinea": "Guinea",

    # Yemen
    "Yemen Arab Rep": "Yemen",
    "Yemen P Dem Rep": "Yemen",
}

# Aplicar reemplazo
df["Country"] = df["Country"].replace(replace_dict)

In [16]:
df["Country"].value_counts().head(20)

Country
United States of America (the)    949
China                             906
India                             658
Philippines (the)                 612
Indonesia                         528
Bangladesh                        317
Japan                             272
Mexico                            262
Viet Nam                          241
Australia                         240
Brazil                            227
Pakistan                          213
Iran (Islamic Republic of)        205
Afghanistan                       200
Colombia                          188
Peru                              178
France                            173
Russian Federation (the)          173
Thailand                          155
Italy                             147
Name: count, dtype: int64

#### COLUMNA ISO – Código ISO del país

- Índice 12
- Ya que aparece incluído en 'Dis No' y podemos averiguarlo, eliminamos.

#### COLUMNA Region – Región (ej. Americas, Asia)

- Índice 13
- Datos de tipo objeto (string)
- No disponemos de valores nulos. Completa la información de "Continente". Dejamos ambas para facilitar las visualizaciones (por si queremos filtrar por Continente o región posteriormente).

In [17]:
# Tabla de frecuencia cruzada
tabla_region_cont = pd.crosstab(df['Region'], df['Continent'])

tabla_region_cont

Continent                  Africa  Americas  Asia  Europe  Oceania
Region                                                            
Australia and New Zealand       0         0     0       0      312
Caribbean                       0       530     0       0        0
Central America                 0       751     0       0        0
Central Asia                    0         0   141       0        0
Eastern Africa               1128         0     0       0        0
Eastern Asia                    0         0  1578       0        0
Eastern Europe                  0         0     0     535        0
Melanesia                       0         0     0       0      232
Micronesia                      0         0     0       0       43
Middle Africa                 421         0     0       0        0
Northern Africa               305         0     0       0        0
Northern America                0      1075     0       0        0
Northern Europe                 0         0     0     207        0
Polynesia                       0         0     0       0       75
Russian Federation              0         0     0      47        0
South America                   0      1163     0       0        0
South-Eastern Asia              0         0  1809       0        0
Southern Africa               200         0     0       0        0
Southern Asia                   0         0  1849       0        0
Southern Europe                 0         0     0     577        0
Western Africa                758         0     0       0        0
Western Asia                    0         0   416       0        0
Western Europe                  0         0     0     492        0

#### COLUMNA Continent - Continente

- Índice 14
- Datos de tipo objeto (string)
- No tenemos valores nulos

#### COLUMNA Location – Lugar específico

- Índice 15
- 1346 nulos. Los modifico reiterando la información de "Region", para no perder información pero eliminar los nulos (similar a lo que hice con Types y Subtypes)

In [18]:
df["Location"] = df["Location"].fillna(df["Region"])

In [19]:
print("Nulos después:", df["Location"].isna().sum())

Nulos después: 0


#### COLUMNA Origin – Origen del desastre (ej. Natural, Industrial)

- Índice 16
- 10864 nulos, he intentado agrupar por categorías pero es imposible ya que agrupa muchos en "Other" debido a typos y textos largos. De momento la dejo por si queremos revisar algún dato concreto.

In [20]:
df["Origin"].unique()

array([nan, 'Typhoon ', 'Heavy rains', 'Landslide', 'Heavy rain',
       'Earthquake and landslide', 'Famine', 'El Nino',
       'Extra-tropical cyclone', 'Brief torrential rain',
       'Heavy rains and thunderstorms', 'Monsoonal rain',
       'Monssonal rain', 'Heavy rains, tremor', 'Brief torrential rains',
       'Monsoon rain', 'Tropical cyclone', 'Drought period',
       'Prolonged drought', 'Monsoon rains',
       'Volcanic activity, tropical rainstorms and ash from the eruptions',
       'monsoonal rain', 'Heavy rain and Snowmelt',
       'Snowmelt and Heavy rains', 'Rain, snowmelt', 'Earthquake',
       'Snowmelt', 'Hevay rain', 'Typhoon Gali', 'Mosoonal rain',
       'Heavy rain and brief torrential rain',
       'Breakdown water supply, poor hygiene practices and increase of number of displaced perdons',
       'Flood', 'Heavy rains, melting snow', 'Continous rainfall',
       'Snow storm and heavy winds', 'Rain and snowmelt', 'Lack rainfall',
       'Food consumed', 'El Nin

#### COLUMNA Associated Dis – Desastres asociados (1)

- Índice 17
- 78% de nulos, intento de rellenar con la info de Associated Dis2 pero tenemos 95% nulos en esa. Aun así, nos falta mucha info (11403). 

#### COLUMNA Associated Dis2 – Desastres asociados (2)

- Índice 18
- 95% de nulos. Tras intentar sacar info para completar la anterior, eliminamos.

In [21]:
df["Associated Dis"] = df["Associated Dis"].fillna(df["Associated Dis2"])

#### COLUMNA OFDA Response – Respuesta de la Office of U.S. Foreign Disaster Assistance

- Índice 19
- 13194 nulos. Eliminamos. 

#### COLUMNA Appeal – Si hubo apelación internacional

- Índice 20
- 12204 nulos. Eliminamos.

#### COLUMNA Declaration – Declaración oficial de desastre.

- Índice 21
- 11517 nulos. Podemos considerar que la mayoría no se consideraron oficialmente, o no disponemos de registros. Eliminamos.

#### COLUMNA Aid Contribution – Contribución de ayuda (si se registró)

- Índice 22
- 95% de nulos y muchos sin registrar. Eliminamos

#### COLUMNA DIS MAG VALUE: Valor de magnitud (ej. magnitud sísmica)

- Índice 23 
- Datos de tipo float que muestran la magnitud asociada
- Encontramos muchos nulos (10.075) pero mantenemos el dato por su importancia.

In [22]:
df['Dis Mag Value'].isnull().sum()

np.int64(10075)

#### COLUMNA DIS MAG SCALE: Escala de magnitud (ej. Richter, Mercalli)

- Índice 24 
- Datos de tipo objeto que representan unidades de medida de distintos tipos de eventos, donde:
    - '°C' (grados centígrados) son Disaster Type = 'Extreme temperature' por Disaster Subtype = 'Cold wave' o 'Heat wave' (584 filas)
    - 'Vaccinated' (vacunados) son Disaster Type = 'Epidemic' por Disaster Subtype = 'Bacterial disease' o 'Viral disease' (1.436 filas)
    - 'Richter' (escala de magnitud de liberación de energía del sismo) son Disaster Type = 'Earthquake' por Disaster Subtype = 'Ground movement' (1.137 filas)
    - 'Kph' (Km por hora, velocidad del viento) son Disaster Type = 'Storm' por Disaster Subtype = 'Tropical cyclone' o 'Convective storm' (4.005 filas)
    - 'Km2' son Disaster Type = 'Flood', por Disaster Subtype = 'Flash flood' o NaN, o 'Drought', por Disaster Subtype = 'Drought' o NaN (6.409 filas)
- Encontramos nulos (1.073 filas) que también representan desastres aunque sin magnitud:
    - Disaster Type = 'Landslide' por Disaster Subtype = 'Avalanche' o 'Landslide' 
    - Disaster Type = 'Mass movement (dry)' por Disaster Subtype = 'Landslide'
    - Disaster Type = 'Volcanic activity' por Disaster Subtype = 'Lava flow' o 'Pyroclastic flow'

In [23]:
df['Dis Mag Scale'].unique()

array(['Km2', 'Kph', nan, 'Richter', 'Vaccinated', '°C'], dtype=object)

#### COLUMNA LATITUDE: Latitud del evento

- Índice 25 
- Datos de tipo objeto (string) con valores positivos (hemisferio norte, de 0° a + 90°) y negativos (hemisferio sur, de 0° a -90°)
- Los datos aparecen con puntos (.) necesarios para que se entiendan como decimales en Tableau
- Encontramos muchos nulos (12.313) pero mantenemos el dato para mapear

In [24]:
df['Latitude'].unique()

array([nan, '24.185', '6.375', ..., '-331.035', '492.688', '132.586'],
      shape=(2092,), dtype=object)

#### COLUMNA LONGITUDE: Longitud del evento

- Índice 26 
- Datos de tipo objeto (string). La longitud se mide desde el meridiano de Greenwich (0°) hacia el este y hacia el oeste, con valores positivos (hemisferio este, de 0° a +180°) y negativos (hemisferio oeste, de 0° a -180°). 
- Los datos aparecen con puntos (.) necesarios para que se entiendan como decimales en Tableau
- Encontramos muchos nulos (12.309) pero mantenemos el dato para mapear

In [25]:
df['Longitude'].unique()

array([nan, '102.543', '-77.438', ..., '-560.461', '23.628', '444.372'],
      shape=(2096,), dtype=object)

#### COLUMNA LOCAL TIME: Hora local del inicio del desastre

- Índice 27 
- Datos de tipo objeto
- Encontramos muchos nulos (13.879) y no es un dato relevante. Eliminamos.

#### COLUMNA RIVER BASIN: Cuenca hidrográfica (si aplica)

- Índice 28 
- Datos de tipo objeto
- Encontramos muchos nulos (13.359) y no es un dato relevante. Eliminamos.

#### COLUMNAS START YEAR (Año de inicio), START MONTH (Mes de inicio) Y START DAY (Día de inicio)

- Índices 29, 30 y 31 
- Datos de tipo integer o float que representan la fecha inicial del suceso pero aparecen en 3 columnas
- Encontramos nulos (valores faltantes) en mes y día

##### SOLUCIÓN:
- Unimos las 3 columnas para generar una única columna de Fecha de inicio "Start Date"
- Seguimos encontrando 3.069 nulos que aparecen como NaT

#### COLUMNAS END YEAR (Año de fin), END MONTH (Mes de fin) Y END DAY (Día de fin)

- Índices 32, 33 y 34 
- Datos de tipo integer o float que representan la fecha final del suceso pero aparecen en 3 columnas
- Encontramos nulos (valores faltantes) en mes y día

##### SOLUCIÓN:
- Unimos las 3 columnas para generar una única columna de Fecha de fin, "End Date"
- Seguimos encontrando 2.999 nulos que aparecen como NaT
- Generamos una columna más, Duración del suceso, "Duration"
- Seguimos encontrando 3.179 nulos que aparecen como NaT

In [26]:
# Función para construir fechas solo si hay año, mes y día
def construir_fecha(year, month, day):
    if pd.notna(year) and pd.notna(month) and pd.notna(day):
        try:
            return pd.Timestamp(int(year), int(month), int(day))
        except:
            return pd.NaT
    return pd.NaT
# NaT (“Not a Time”) → es la versión especial de NaN, pero para valores de tipo fecha/hora

# Crear columnas de fecha de inicio y fin
df["Start Date"] = df.apply(lambda row: construir_fecha(row["Start Year"], row["Start Month"], row["Start Day"]), axis=1)
df["End Date"] = df.apply(lambda row: construir_fecha(row["End Year"], row["End Month"], row["End Day"]), axis=1)

In [27]:
# Fechas de inicio:
df["Start Date"]

0       1970-01-04
1              NaT
2              NaT
3       1970-11-12
4       1970-04-13
           ...    
14639   2021-07-16
14640   2021-02-01
14641   2021-09-07
14642   2021-01-11
14643          NaT
Name: Start Date, Length: 14644, dtype: datetime64[ns]

In [28]:
# Fechas de fin:
df["End Date"] 

0       1970-01-04
1              NaT
2              NaT
3       1970-11-12
4       1970-04-13
           ...    
14639   2021-08-07
14640   2021-02-15
14641   2021-09-13
14642   2021-01-12
14643   2021-10-07
Name: End Date, Length: 14644, dtype: datetime64[ns]

In [29]:
# Sacamos la Duración del suceso:
df["Duration"] = (df["End Date"] - df["Start Date"]).dt.days + 1
df["Duration"]
# .dt.days convierte el Timedelta en número de días (entero)

0         1.0
1         NaN
2         NaN
3         1.0
4         1.0
         ... 
14639    23.0
14640    15.0
14641     7.0
14642     2.0
14643     NaN
Name: Duration, Length: 14644, dtype: float64

#### COLUMNA TOTAL DEATHS (Número total de fallecidos)

- Índices 35  
- Datos de tipo float 
- Encontramos 4.445 nulos

In [30]:
df["Total Deaths"].isnull().sum()

np.int64(4445)

#### COLUMNAS Nº INJURED (Número de heridos), Nº AFFECTED (Número de afectados), Nº HOMELESS (Número de personas sin hogar) Y TOTAL AFFECTED (Total de personas afectadas (suma de afectados, heridos, sin hogar))

- Índices 36, 37, 38 y 39 
- Datos de tipo float todos relacionados con personas afectadas por los desastres naturales
- Encontramos nulos:
    - No Injured: 10.993 nulos
    - No Affected: 5.798 nulos
    - No Homeless: 12.395 nulos
    - Total Affected: 3.603 nulos
- Siendo "Total Affected" la suma de los otros 3, nos quedamos sólo con éste 

In [31]:
# Comprobamos si "Total Affected" es la suma de los 3, pero sale False cuando hay NaN: 
suma = df[["No Injured", "No Affected", "No Homeless"]].fillna(0).sum(axis=1)
df["check"] = suma == df["Total Affected"]
df["check"]

0        False
1        False
2        False
3         True
4         True
         ...  
14639     True
14640     True
14641     True
14642     True
14643     True
Name: check, Length: 14644, dtype: bool

#### COLUMNAS RECONSTRUCTION COSTS (Costes de reconstrucción, en miles de USD), INSURED DAMAGES (Daños asegurados, en miles de USD), TOTAL DAMAGES (Daños totales, en miles de USD) Y CPI (Índice de precios al consumidor)

- Índices 40, 41, 42 y 43 
- Datos de tipo float
- Encontramos nulos:
    - Reconstruction Costs ('000 US$): 100% nulos
    - Insured Damages ('000 US$): 95% nulos
    - Total Damages ('000 US$): 65% nulos
    - CPI: 315 nulos
- Eliminamos las 4 por la cantidad de nulos de las 3 primeras, y por no tener interés en el proyecto el CPI

COLUMNA Adm Level – Nivel administrativo (ej. nacional, regional)

- Índice 44
- 6789 nulos. La eliminamos ya que utilizaremos las otras columnas de localización.

In [32]:
df["Adm Level"].isna().sum()

np.int64(6789)

COLUMNA Admin1 Code – Código de primera división administrativa (ej. estado, provincia)

- Índice 45
- 10063 nulos. La eliminamos.

In [33]:
df["Admin1 Code"].isna().sum()

np.int64(10063)

COLUMNA Admin2 Code – Código de segunda división administrativa (ej. municipio)

- Índice 46
- 10679 nulos. La eliminamos.


In [34]:
df["Admin2 Code"].isna().sum()

np.int64(10679)

COLUMNA Geo Locations – Ubicaciones geográficas adicionales

- Índice 47
- La eliminamos ya que reitera la información de Location.

-------------------

COLUMNAS QUE ELIMINAMOS

In [35]:
# COLUMNAS AÑADIDAS:
["Duration", "End Date", "Start Date"]

#COLUMNAS ELIMINADAS:
df_revisado=df.drop(columns=['Seq', 'Glide', 'Disaster Group', 'Disaster Subsubtype', 'ISO', 'Associated Dis2','OFDA Response', 'Appeal', 'Declaration', 'Aid Contribution', 'Local Time', 'River Basin', 'Start Year', 'Start Month', 'Start Day', 'End Year', 'End Month', 'End Day', 'No Injured', 'No Affected', 'No Homeless', 'check', "Reconstruction Costs ('000 US$)", "Insured Damages ('000 US$)", "Total Damages ('000 US$)", 'CPI', 'Adm Level', 'Admin1 Code', 'Admin2 Code', 'Geo Locations'])


In [36]:
df_revisado.sample(5)

Dis No  Year Disaster Subgroup Disaster Type   Disaster Subtype  \
13584  2018-0204-VNM  2018      Hydrological         Flood        Flash flood   
13551  2017-0511-USA  2017    Climatological      Wildfire           Wildfire   
14031  2019-0397-NGA  2019        Biological      Epidemic  Bacterial disease   
1686   1983-0153-IND  1983      Hydrological         Flood              Flood   
4434   1996-0221-KHM  1996      Hydrological         Flood     Riverine flood   

      Event Name                         Country              Region  \
13584        NaN                        Viet Nam  South-Eastern Asia   
13551   'Thomas'  United States of America (the)    Northern America   
14031    Cholera                         Nigeria      Western Africa   
1686         NaN                           India       Southern Asia   
4434         NaN                        Cambodia  South-Eastern Asia   

      Continent                                           Location  \
13584      Asia                                           Lai Chau   
13551  Americas  Ventura, Santa Barbara, Los Angeles, San Diego...   
14031    Africa                                      Adamawa State   
1686       Asia                                         Tamil Nadu   
4434       Asia  Kratie, Phnom Penh, Tatanakiri, Stung Treng, K...   

              Origin                 Associated Dis  Dis Mag Value  \
13584            NaN  Slide (land, mud, snow, rock)            NaN   
13551  Droght, winds                            NaN         1102.0   
14031            NaN                            NaN            NaN   
1686             NaN                            NaN            NaN   
4434             NaN                            NaN         1306.0   

      Dis Mag Scale Latitude Longitude  Total Deaths  Total Affected  \
13584           Km2      NaN       NaN          33.0         16728.0   
13551           Km2      NaN       NaN           2.0          2963.0   
14031    Vaccinated      NaN       NaN           4.0           665.0   
1686            Km2      NaN       NaN          58.0             NaN   
4434            Km2      NaN       NaN          59.0       1300000.0   

      Start Date   End Date  Duration  
13584 2018-06-23 2018-06-27       5.0  
13551 2017-12-04 2017-12-31      28.0  
14031        NaT 2019-08-21       NaN  
1686         NaT        NaT       NaN  
4434  1996-09-30 1996-11-03      35.0

In [37]:
# Estas 3 columnas aparecen con decimales cuando son deben ser números enteros. Las modificamos:
df_revisado["Total Deaths"] = df_revisado["Total Deaths"].astype("Int64")
df_revisado["Total Affected"] = df_revisado["Total Affected"].astype("Int64")
df_revisado["Duration"] = df_revisado["Duration"].astype("Int64")

In [38]:
df_revisado.sample(5)

Dis No  Year Disaster Subgroup         Disaster Type  \
12029  2011-0663-JPN  2011      Hydrological                 Flood   
7632   2004-0652-LKA  2004      Hydrological                 Flood   
11853  2015-0171-COL  2015      Hydrological             Landslide   
12552  2015-0177-USA  2015    Meteorological                 Storm   
5048   1998-0212-USA  1998    Meteorological  Extreme temperature    

       Disaster Subtype Event Name                         Country  \
12029             Flood        NaN                           Japan   
7632     Riverine flood        NaN                       Sri Lanka   
11853         Landslide        NaN                        Colombia   
12552  Convective storm        NaN  United States of America (the)   
5048          Heat wave        NaN  United States of America (the)   

                 Region Continent  \
12029      Eastern Asia      Asia   
7632      Southern Asia      Asia   
11853     South America  Americas   
12552  Northern America  Americas   
5048   Northern America  Americas   

                                                Location          Origin  \
12029  Yasudamati, Kamosi districts (Niigata province...     Heavy rains   
7632   Kilinochchi, Jaffna, Vavuniya, Mullattivu dist...  Monsoonal rain   
11853         Salgar city (Antioquia), Amazonas province     Heavy rains   
12552  Nashville town (Howard district, Arkansas prov...             NaN   
5048     Arizona, Florida, Colorado, Texas, Oklahoma, LA             NaN   

              Associated Dis  Dis Mag Value Dis Mag Scale Latitude Longitude  \
12029  Broken Dam/Burst bank            NaN           Km2      NaN       NaN   
7632                     NaN        23460.0           Km2      NaN       NaN   
11853                  Flood            NaN           NaN  5.70693     -74.5   
12552                    NaN            NaN           Kph      NaN       NaN   
5048                 Drought           38.0            °C      NaN       NaN   

       Total Deaths  Total Affected Start Date   End Date  Duration  
12029             6            2000 2011-07-29 2011-07-31         3  
7632              6          200000 2004-12-11 2004-12-23        13  
11853            93             348 2015-05-17 2015-05-18         2  
12552             6            <NA> 2015-05-06 2015-05-13         8  
5048            130            <NA> 1998-05-01 1998-07-20        81

In [39]:
df_revisado.to_csv("datos_revisados.csv")

In [40]:
import pandas as pd
import re

def corregir_coordenadas(coord):
    """
    Convierte una coordenada en formato de texto (ej. '30.37 N', '78.30 W')
    a un formato numérico decimal.
    - 'N' y 'E' son positivos.
    - 'S' y 'W' son negativos.
    """
    # Si la coordenada ya es un número o está vacía (NaN), no se hace nada.
    if pd.isna(coord) or isinstance(coord, (int, float)):
        return coord

    # Asegurarse de que la coordenada es una cadena de texto para procesarla.
    coord_str = str(coord).strip().upper()

    # Usar expresiones regulares para encontrar el número.
    numero_match = re.search(r'[-+]?\d*\.\d+|\d+', coord_str)
    if not numero_match:
        return None # Retorna None si no se encuentra un número válido.

    numero = float(numero_match.group(0))

    # Aplicar el signo correcto según la letra (N, S, E, W).
    if 'S' in coord_str or 'W' in coord_str:
        return -abs(numero)
    
    return abs(numero)

try:
    # 1. Cargar el archivo CSV en un DataFrame de pandas
    # Asegúrate de que la ruta del archivo sea la correcta.
    ruta_archivo = r'c:\Users\krigu\OneDrive\Desktop\ADALAB\Modulo-4\Proyecto_desastres_naturales_modulo_4\datos_revisados.csv'
    df = pd.read_csv(ruta_archivo)

    # 2. Crear las nuevas columnas aplicando la función de corrección
    print("Corrigiendo las columnas 'Latitude' y 'Longitude'...")
    df['Latitud_corregida'] = df['Latitude'].apply(corregir_coordenadas)
    df['Longitud_corregida'] = df['Longitude'].apply(corregir_coordenadas)
    print("Corrección completada.")

    # 3. Mostrar una vista previa de las columnas originales y las corregidas
    print("\nVista previa de los datos corregidos:")
    print(df[['Latitude', 'Latitud_corregida', 'Longitude', 'Longitud_corregida']].head(10))
    
    # Ejemplo de una fila que necesitaba corrección (fila 206)
    print("\nEjemplo de una fila específica (índice 206):")
    print(df.loc[206, ['Latitude', 'Latitud_corregida', 'Longitude', 'Longitud_corregida']])

    # 4. Guardar el DataFrame con las nuevas columnas en un nuevo archivo CSV
    ruta_salida = r'c:\Users\krigu\OneDrive\Desktop\ADALAB\Modulo-4\Proyecto_desastres_naturales_modulo_4\datos_corregidos.csv'
    df.to_csv(ruta_salida, index=False)
    print(f"\n¡Éxito! El archivo con las coordenadas corregidas se ha guardado en: {ruta_salida}")

except FileNotFoundError:
    print(f"Error: No se pudo encontrar el archivo en la ruta especificada.")
except Exception as e:
    print(f"Ha ocurrido un error inesperado: {e}")


Corrigiendo las columnas 'Latitude' y 'Longitude'...
Corrección completada.

Vista previa de los datos corregidos:
  Latitude  Latitud_corregida Longitude  Longitud_corregida
0      NaN                NaN       NaN                 NaN
1      NaN                NaN       NaN                 NaN
2      NaN                NaN       NaN                 NaN
3      NaN                NaN       NaN                 NaN
4      NaN                NaN       NaN                 NaN
5      NaN                NaN       NaN                 NaN
6      NaN                NaN       NaN                 NaN
7      NaN                NaN       NaN                 NaN
8      NaN                NaN       NaN                 NaN
9      NaN                NaN       NaN                 NaN

Ejemplo de una fila específica (índice 206):
Latitude               30.37 N
Latitud_corregida        30.37
Longitude             104.06 E
Longitud_corregida      104.06
Name: 206, dtype: object

¡Éxito! El archivo con las co

In [41]:
import pandas as pd
import unicodedata
import re

def limpiar_titulos(df):
    """
    Limpia y estandariza los títulos de las columnas de un DataFrame.
    - Convierte a minúsculas.
    - Elimina acentos y caracteres especiales.
    - Reemplaza espacios y caracteres no alfanuméricos por guiones bajos.
    """
    nuevas_columnas = []
    for col in df.columns:
        # 1. Normalizar para descomponer acentos (ej. 'á' -> 'a' + '´')
        col_normalizada = unicodedata.normalize('NFKD', col)
        # 2. Codificar a ASCII ignorando los caracteres no ASCII (los acentos)
        col_ascii = col_normalizada.encode('ascii', 'ignore').decode('utf-8')
        # 3. Convertir a minúsculas y reemplazar espacios/caracteres no alfanuméricos
        col_limpia = re.sub(r'[^a-zA-Z0-9]+', '_', col_ascii).lower()
        # 4. Eliminar guiones bajos al principio o al final
        col_limpia = col_limpia.strip('_')
        nuevas_columnas.append(col_limpia)
    
    df.columns = nuevas_columnas
    return df

try:
    # --- Carga de datos ---
    ruta_archivo = r'c:\Users\krigu\OneDrive\Desktop\ADALAB\Modulo-4\Proyecto_desastres_naturales_modulo_4\datos_corregidos.csv'
    df = pd.read_csv(ruta_archivo)

    # --- 1. Rellenar valores nulos en las coordenadas corregidas ---
    print("Rellenando valores nulos en 'Latitud_corregida' y 'Longitud_corregida'...")
    
    # Usamos .fillna() que es la forma más eficiente en pandas para esta tarea.
    # Rellena los nulos en 'Latitud_corregida' con los valores de 'Latitude'.
    df['Latitud_corregida'].fillna(df['Latitude'], inplace=True)
    
    # Rellena los nulos en 'Longitud_corregida' con los valores de 'Longitude'.
    df['Longitud_corregida'].fillna(df['Longitude'], inplace=True)
    
    print("Valores nulos rellenados.")

    # --- 2. Limpieza de los títulos de las columnas ---
    print("\nLimpiando los títulos de las columnas para mejor compatibilidad...")
    
    # Eliminar la columna 'Unnamed: 0' si existe
    if 'Unnamed: 0' in df.columns:
        df = df.drop(columns=['Unnamed: 0'])
        print("Columna 'Unnamed: 0' eliminada.")
        
    # Aplicar la función de limpieza a todos los títulos
    df = limpiar_titulos(df)
    print("Títulos de columnas estandarizados.")
    print("\nNuevos títulos de las columnas:")
    print(df.columns.tolist())

    # --- Verificación y guardado ---
    print("\nVista previa de las columnas de coordenadas actualizadas:")
    # Mostramos las 4 columnas de coordenadas para verificar el relleno de nulos
    print(df[['latitude', 'latitud_corregida', 'longitude', 'longitud_corregida']].head(10))

    # Guardar el DataFrame final en un nuevo archivo CSV
    ruta_salida = r'c:\Users\krigu\OneDrive\Desktop\ADALAB\Modulo-4\Proyecto_desastres_naturales_modulo_4\datos_finales_limpios.csv'
    df.to_csv(ruta_salida, index=False)
    
    print(f"\n¡Proceso completado! El archivo final ha sido guardado en:\n{ruta_salida}")

except FileNotFoundError:
    print(f"Error: No se pudo encontrar el archivo en la ruta: {ruta_archivo}")
except Exception as e:
    print(f"Ha ocurrido un error inesperado: {e}")

Rellenando valores nulos en 'Latitud_corregida' y 'Longitud_corregida'...
Valores nulos rellenados.

Limpiando los títulos de las columnas para mejor compatibilidad...
Columna 'Unnamed: 0' eliminada.
Títulos de columnas estandarizados.

Nuevos títulos de las columnas:
['dis_no', 'year', 'disaster_subgroup', 'disaster_type', 'disaster_subtype', 'event_name', 'country', 'region', 'continent', 'location', 'origin', 'associated_dis', 'dis_mag_value', 'dis_mag_scale', 'latitude', 'longitude', 'total_deaths', 'total_affected', 'start_date', 'end_date', 'duration', 'latitud_corregida', 'longitud_corregida']

Vista previa de las columnas de coordenadas actualizadas:
  latitude  latitud_corregida longitude  longitud_corregida
0      NaN                NaN       NaN                 NaN
1      NaN                NaN       NaN                 NaN
2      NaN                NaN       NaN                 NaN
3      NaN                NaN       NaN                 NaN
4      NaN                NaN    

C:\Users\krigu\AppData\Local\Temp\ipykernel_9068\2170880457.py:37: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['Latitud_corregida'].fillna(df['Latitude'], inplace=True)
C:\Users\krigu\AppData\Local\Temp\ipykernel_9068\2170880457.py:37: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '[nan nan nan ... nan nan nan]' has dtype incompatible with float64, please explicitly cast to a compatible dtype first.
  df['Latitud_corregida'].fillna(df['Latit

In [42]:
df_finales = pd.read_csv("datos_finales_limpios.csv", index_col=0)
df_finales = df_finales.reset_index()
df_finales.head()

dis_no  year disaster_subgroup disaster_type  disaster_subtype  \
0  1970-0013-ARG  1970      Hydrological         Flood             Flood   
1  1970-0109-AUS  1970    Meteorological         Storm  Tropical cyclone   
2  1970-0044-BEN  1970      Hydrological         Flood             Flood   
3  1970-0063-BGD  1970    Meteorological         Storm  Tropical cyclone   
4  1970-0026-BGD  1970    Meteorological         Storm             Storm   

  event_name     country                     region continent  \
0        NaN   Argentina              South America  Americas   
1        Ada   Australia  Australia and New Zealand   Oceania   
2        NaN       Benin             Western Africa    Africa   
3        NaN  Bangladesh              Southern Asia      Asia   
4        NaN  Bangladesh              Southern Asia      Asia   

             location origin associated_dis  dis_mag_value dis_mag_scale  \
0             Mendoza    NaN            NaN            NaN           Km2   
1          Queensland    NaN            NaN            NaN           Kph   
2      Atacora region    NaN            NaN            NaN           Km2   
3  Khulna, Chittagong    NaN            NaN            NaN           Kph   
4       Southern Asia    NaN            NaN            NaN           Kph   

  latitude longitude  total_deaths  total_affected  start_date    end_date  \
0      NaN       NaN          36.0             NaN  1970-01-04  1970-01-04   
1      NaN       NaN          13.0             NaN         NaN         NaN   
2      NaN       NaN           NaN             NaN         NaN         NaN   
3      NaN       NaN      300000.0       3648000.0  1970-11-12  1970-11-12   
4      NaN       NaN          17.0           110.0  1970-04-13  1970-04-13   

   duration  latitud_corregida  longitud_corregida  
0       1.0                NaN                 NaN  
1       NaN                NaN                 NaN  
2       NaN                NaN                 NaN  
3       1.0                NaN                 NaN  
4       1.0                NaN                 NaN

In [ ]:
# pip install geopy


Note: you may need to restart the kernel to use updated packages.


In [44]:
import pandas as pd
from geopy.geocoders import Nominatim
import time

# --- Inicializar el geocodificador y un caché para las coordenadas ---
# El user_agent es un nombre único para tu aplicación.
geolocator = Nominatim(user_agent="corrector_desastres_v2")
location_cache = {}

def obtener_coordenadas(location_str, country_str):
    """
    Busca las coordenadas de una ubicación. Intenta primero con la localización
    específica y, si falla, usa el país como respaldo.
    """
    # Clave para el caché: combina localización y país para ser más preciso.
    cache_key = f"{location_str}, {country_str}".strip().lower()
    if cache_key in location_cache:
        return location_cache[cache_key]

    coords = None
    
    # 1. Intentar con la localización específica (ej. "Mendoza, Argentina")
    if pd.notna(location_str):
        try:
            full_location = f"{location_str}, {country_str}"
            # Pausa para no sobrecargar el servicio gratuito de geocodificación.
            time.sleep(1) 
            location_data = geolocator.geocode(full_location, language='es', timeout=10)
            if location_data:
                coords = (location_data.latitude, location_data.longitude)
        except Exception as e:
            print(f"  -> Advertencia al geocodificar '{full_location}': {e}")

    # 2. Si falla, intentar solo con el país como respaldo.
    if coords is None and pd.notna(country_str):
        print(f"  -> No se encontró '{location_str}'. Intentando con el país: '{country_str}'")
        try:
            time.sleep(1)
            location_data = geolocator.geocode(country_str, language='es', timeout=10)
            if location_data:
                coords = (location_data.latitude, location_data.longitude)
        except Exception as e:
            print(f"  -> Advertencia al geocodificar '{country_str}': {e}")

    # Guardar en caché el resultado (incluso si es None) para no repetir la búsqueda.
    location_cache[cache_key] = coords
    return coords

try:
    # --- Carga de datos ---
    # Usamos el archivo que ya tiene los títulos limpios.
    ruta_archivo = r'c:\Users\krigu\OneDrive\Desktop\ADALAB\Modulo-4\Proyecto_desastres_naturales_modulo_4\datos_finales_limpios.csv'
    df = pd.read_csv(ruta_archivo)

    print("Iniciando la validación de coordenadas usando la columna 'location'...")
    
    filas_corregidas = 0
    # Iteramos sobre las filas que tienen coordenadas potencialmente inválidas para optimizar.
    # Un valor es inválido si no es un número o si está fuera del rango permitido.
    df_invalidos = df[
        (pd.to_numeric(df['latitud_corregida'], errors='coerce').isna()) |
        (df['latitud_corregida'].abs() > 90) |
        (pd.to_numeric(df['longitud_corregida'], errors='coerce').isna()) |
        (df['longitud_corregida'].abs() > 180)
    ]

    print(f"Se encontraron {len(df_invalidos)} filas con coordenadas potencialmente inválidas a revisar.")

    for index, row in df_invalidos.iterrows():
        lat = row['latitud_corregida']
        lon = row['longitud_corregida']
        
        # Comprobar si la latitud o longitud son inválidas
        lat_invalida = pd.notna(lat) and (lat > 90 or lat < -90)
        lon_invalida = pd.notna(lon) and (lon > 180 or lon < -180)

        # Solo proceder si realmente hay un error
        if lat_invalida or lon_invalida:
            location = row['location']
            country = row['country']
            
            print(f"\nFila {index}: Coordenadas inválidas (Lat: {lat}, Lon: {lon}).")
            print(f" -> Buscando corrección para: '{location}', '{country}'...")
            
            # Obtener coordenadas usando la nueva lógica (location > country)
            coords_nuevas = obtener_coordenadas(location, country)
            
            if coords_nuevas:
                lat_nueva, lon_nueva = coords_nuevas
                # Actualizar las columnas en el DataFrame original
                df.loc[index, 'latitud_corregida'] = lat_nueva
                df.loc[index, 'longitud_corregida'] = lon_nueva
                print(f" -> ¡Corregido! Nuevas coordenadas: (Lat: {lat_nueva:.4f}, Lon: {lon_nueva:.4f})")
                filas_corregidas += 1
            else:
                print(f" -> No se pudo encontrar una coordenada válida para '{location}, {country}'. Se mantendrá el valor original.")

    if filas_corregidas == 0:
        print("\nNo se encontraron coordenadas fuera de los rangos válidos que necesitaran corrección.")
    else:
        print(f"\nSe corrigieron un total de {filas_corregidas} filas con coordenadas inválidas.")

    # --- Guardado final ---
    ruta_salida = r'c:\Users\krigu\OneDrive\Desktop\ADALAB\Modulo-4\Proyecto_desastres_naturales_modulo_4\datos_finales_corregidos_v2.csv'
    df.to_csv(ruta_salida, index=False)
    
    print(f"\n¡Proceso finalizado! El archivo con las coordenadas corregidas y validadas se ha guardado en:\n{ruta_salida}")

except FileNotFoundError:
    print(f"Error: No se pudo encontrar el archivo en la ruta: {ruta_archivo}")
    print("Asegúrate de haber ejecutado el script anterior para generar 'datos_finales_limpios.csv'.")
except Exception as e:
    print(f"Ha ocurrido un error inesperado: {e}")


Iniciando la validación de coordenadas usando la columna 'location'...
Se encontraron 12381 filas con coordenadas potencialmente inválidas a revisar.

Fila 1275: Coordenadas inválidas (Lat: 24.688, Lon: 991.57).
 -> Buscando corrección para: 'Syhlet', 'Bangladesh'...
  -> No se encontró 'Syhlet'. Intentando con el país: 'Bangladesh'
 -> ¡Corregido! Nuevas coordenadas: (Lat: 24.4769, Lon: 90.2934)

Fila 2057: Coordenadas inválidas (Lat: 26.755, Lon: 616.0).
 -> Buscando corrección para: 'Dharbhanga, Madhubani, Saharsa, Munger, Khagana, Bihar Sharie (Bihar State)', 'India'...
  -> No se encontró 'Dharbhanga, Madhubani, Saharsa, Munger, Khagana, Bihar Sharie (Bihar State)'. Intentando con el país: 'India'
 -> ¡Corregido! Nuevas coordenadas: (Lat: 22.3511, Lon: 78.6677)

Fila 10616: Coordenadas inválidas (Lat: 125.091, Lon: 259.63).
 -> Buscando corrección para: 'Bohol district (Central Visayas province), Biliran, Eastern Samar, Leyte, Northern Samar, Samar, Southern Leyte districts (Easte

In [46]:
df_finales_corregidos = pd.read_csv("datos_finales_corregidos_v2.csv", index_col=0)
df_finales_corregidos = df_finales_corregidos.reset_index()
df_finales_corregidos.head()

dis_no  year disaster_subgroup disaster_type  disaster_subtype  \
0  1970-0013-ARG  1970      Hydrological         Flood             Flood   
1  1970-0109-AUS  1970    Meteorological         Storm  Tropical cyclone   
2  1970-0044-BEN  1970      Hydrological         Flood             Flood   
3  1970-0063-BGD  1970    Meteorological         Storm  Tropical cyclone   
4  1970-0026-BGD  1970    Meteorological         Storm             Storm   

  event_name     country                     region continent  \
0        NaN   Argentina              South America  Americas   
1        Ada   Australia  Australia and New Zealand   Oceania   
2        NaN       Benin             Western Africa    Africa   
3        NaN  Bangladesh              Southern Asia      Asia   
4        NaN  Bangladesh              Southern Asia      Asia   

             location origin associated_dis  dis_mag_value dis_mag_scale  \
0             Mendoza    NaN            NaN            NaN           Km2   
1          Queensland    NaN            NaN            NaN           Kph   
2      Atacora region    NaN            NaN            NaN           Km2   
3  Khulna, Chittagong    NaN            NaN            NaN           Kph   
4       Southern Asia    NaN            NaN            NaN           Kph   

  latitude longitude  total_deaths  total_affected  start_date    end_date  \
0      NaN       NaN          36.0             NaN  1970-01-04  1970-01-04   
1      NaN       NaN          13.0             NaN         NaN         NaN   
2      NaN       NaN           NaN             NaN         NaN         NaN   
3      NaN       NaN      300000.0       3648000.0  1970-11-12  1970-11-12   
4      NaN       NaN          17.0           110.0  1970-04-13  1970-04-13   

   duration  latitud_corregida  longitud_corregida  
0       1.0                NaN                 NaN  
1       NaN                NaN                 NaN  
2       NaN                NaN                 NaN  
3       1.0                NaN                 NaN  
4       1.0                NaN                 NaN

In [48]:
# Estas 3 columnas aparecen con decimales cuando son deben ser números enteros. Las modificamos:
df_finales_corregidos["total_deaths"] = df_finales_corregidos["total_deaths"].astype("Int64")
df_finales_corregidos["total_affected"] = df_finales_corregidos["total_affected"].astype("Int64")
df_finales_corregidos["duration"] = df_finales_corregidos["duration"].astype("Int64")

In [49]:
df_finales_corregidos.head()

dis_no  year disaster_subgroup disaster_type  disaster_subtype  \
0  1970-0013-ARG  1970      Hydrological         Flood             Flood   
1  1970-0109-AUS  1970    Meteorological         Storm  Tropical cyclone   
2  1970-0044-BEN  1970      Hydrological         Flood             Flood   
3  1970-0063-BGD  1970    Meteorological         Storm  Tropical cyclone   
4  1970-0026-BGD  1970    Meteorological         Storm             Storm   

  event_name     country                     region continent  \
0        NaN   Argentina              South America  Americas   
1        Ada   Australia  Australia and New Zealand   Oceania   
2        NaN       Benin             Western Africa    Africa   
3        NaN  Bangladesh              Southern Asia      Asia   
4        NaN  Bangladesh              Southern Asia      Asia   

             location origin associated_dis  dis_mag_value dis_mag_scale  \
0             Mendoza    NaN            NaN            NaN           Km2   
1          Queensland    NaN            NaN            NaN           Kph   
2      Atacora region    NaN            NaN            NaN           Km2   
3  Khulna, Chittagong    NaN            NaN            NaN           Kph   
4       Southern Asia    NaN            NaN            NaN           Kph   

  latitude longitude  total_deaths  total_affected  start_date    end_date  \
0      NaN       NaN            36            <NA>  1970-01-04  1970-01-04   
1      NaN       NaN            13            <NA>         NaN         NaN   
2      NaN       NaN          <NA>            <NA>         NaN         NaN   
3      NaN       NaN        300000         3648000  1970-11-12  1970-11-12   
4      NaN       NaN            17             110  1970-04-13  1970-04-13   

   duration  latitud_corregida  longitud_corregida  
0         1                NaN                 NaN  
1      <NA>                NaN                 NaN  
2      <NA>                NaN                 NaN  
3         1                NaN                 NaN  
4         1                NaN                 NaN

In [51]:
df_finales_corregidos["longitud_corregida"].info()

<class 'pandas.core.series.Series'>
RangeIndex: 14644 entries, 0 to 14643
Series name: longitud_corregida
Non-Null Count  Dtype  
--------------  -----  
2335 non-null   float64
dtypes: float64(1)
memory usage: 114.5 KB


In [52]:
df_finales_corregidos.to_csv("datos_finales_mapas.csv")